Added support for custom JS module path, issue #29

datvm opened the pull request:
This PR improves previous PR to fix #29 . Here're some clarification:

## From end-developer perspective:

All `Create` and `CreateAsync` methods are marked `Obsolete` so they should move on from them and use the new methods from `IFileSystemAccessService` instead (for example, `FileSystemFileHandle.Create` becomes `FileSystemAccessService.CreateFileHandle`).

The obsolete methods still work for backward-compatibility but the custom-path feature (and future options) won't work due to missing options reference.

They can also configure the service directly when adding FSA to the `ServiceCollection`:

```cs
builder.Services.AddFileSystemAccessServiceInProcess(options =>
{
    // The file at this path in this example is manually copied to wwwroot folder
    options.ScriptPath = $"/content/custom-path/{FileSystemAccessOptions.DefaultNamespace}.js";
});
```

## Changes for internal code:

Since a few places still use `Create` and `CreateAsync`, I feel it's harder to keep references to the Service instance. Instead those static method now becomes `internal` and the service simply invokes them with `Option` instance. This serves 2 purposes: there is no code repetition and we do not need to keep a service class reference on each handler instance.

Right now all old (marked `Obsolete`) methods simply point to the newly created `internal` methods for reusability though without `Options` instance certain feature like custom path or anything added in the future won't work.

KristofferStrube commented on src/KristofferStrube.Blazor.FileSystemAccess/FileSystemAccessService.cs:
```diff
@@ -151,6 +159,46 @@ public async Task<bool> IsSupportedAsync()
             await jSRuntime.InvokeAsync<bool>("window.hasOwnProperty", "showDirectoryPicker");
     }
 
+    #region Create Handle Instances
```
For the following methods, we can use the jSRuntime member instead of taking it as an argument. This would be one of the pros of moving the creator methods to the service.

KristofferStrube commented on src/KristofferStrube.Blazor.FileSystemAccess/FileSystemWritableFileStream.InProcess.cs:
```diff
@@ -13,7 +13,7 @@ public class FileSystemWritableFileStreamInProcess : FileSystemWritableFileStrea
 
     public static async Task<FileSystemWritableFileStreamInProcess> CreateAsync(IJSRuntime jSRuntime, IJSInProcessObjectReference jSReference)
```
We should add a creator method to the `IFileSystemAccessService` for the `FileSystemWritableFileStreamInProcess` as we have done with the handles and obsolete this as we have done for the others.

KristofferStrube commented on src/KristofferStrube.Blazor.FileSystemAccess/Options/FileSystemAccessOptions.cs:
```diff
@@ -0,0 +1,12 @@
+    public static readonly string DefaultNamespace = Assembly.GetExecutingAssembly().GetName().Name ?? "KristofferStrube.Blazor.FileSystemAccess";
+    public static readonly string DefaultPath = $"./_content/{DefaultNamespace}/{DefaultNamespace}.js";
+
+    public string ScriptPath { get; set; } = DefaultPath;
```
We will need the base path for the future when creating objects from the `Blazor.FileAPI` and `Blazor.Streams` packages. By the base, I mean the `"./_content/"` path. So I suggest changing the option to also contain a BasePath property that defaults to `"./_content/"` from a static `DefaultBasePath` field and then changing the name of `DefaultPath` to `DefaultScriptPath` with just the value `"{DefaultNamespace}/{DefaultNamespace}.js"`.
This will of cause mean that we will need to concat `BasePath` and `ScriptPath` the places we use it.

KristofferStrube commented on samples/KristofferStrube.Blazor.FileSystemAccess.WasmExample/Program.cs:
```diff
@@ -10,7 +11,14 @@
+builder.Services.AddFileSystemAccessServiceInProcess(options =>
+{
+    // The file at this path in this example is manually copied to wwwroot folder
+    options.ScriptPath = $"/content/custom-path/{FileSystemAccessOptions.DefaultNamespace}.js";
```
I see the point in showing the example of using the option, but I don't think most will do this. So I would add this manually copied script. You can keep this line, but comment it out so that people see that this is possible.

KristofferStrube reviewed (changes requested):
I have left some comments that primarily focus on naming. Awesome job! :D

KristofferStrube commented:
After some more thought, I'm a bit conflicted on the part about changing the position of the creator methods.
I am currently working on homogenizing my pattern for wrapping across all my wrapper projects i.e. `Blazor.Popper`, `Blazor.Streams`, `Blazor.FileAPI`, `Blazor.DeviceOrientation`, and `Blazor.ServiceWorker`.
Some of these projects have no service like `Blazor.Streams` or have a service, but not one that is central to the library like `IURLService` in `Blazor.FileAPI`. If we were to make this change of moving the creator methods then we would need to add services to those projects as well with the only purpose of creating wrapper objects and we would then need to have instances of these services everywhere we need to create objects from those packages which feel too complex.

So I would prefer if we don't add those creator methods in the service and instead make the new internal creator methods on the individual wrapper classes public. So we should also remove the obsolete attributes.

datvm commented on src/KristofferStrube.Blazor.FileSystemAccess/Options/FileSystemAccessOptions.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using System.Reflection;
+
+namespace KristofferStrube.Blazor.FileSystemAccess.Options;
```
This is a good idea. In combination with the other suggestion of staying with `Create` methods, user should be able to put the option instance in without additional namespace.

datvm commented:
@KristofferStrube  I am conflicted about that too. Strictly speaking I chose the current approach just because we can easily use DI for it (without Service we cannot really have access to DI without users passing it to us). Or, we can go with `JsonSerializer` approach. Basically user can set the settings individually with each call _and_ optionally set a global (static) default settings.

datvm commented:
I have pushed 2 commits fixing issues mentioned in your comment. I also cleaned up the Service class a bit because there was too many repetitive codes that make adding the new parameter tedious. You may need to check:

- Any reason `FileSystemAccessServiceInProcess` have to extends `FileSystemAccessService`? Given the code I don't see why you can't just separate the two. For now I simply explicitly implement `IFileSystemAccessService` for `FileSystemAccessServiceInProcess` and it contains a lot of boilerplate code.

```cs
        return serviceCollection
            .AddScoped<IFileSystemAccessServiceInProcess, FileSystemAccessServiceInProcess>()
            .AddScoped(sp =>
            {
                var service = sp.GetRequiredService<IFileSystemAccessServiceInProcess>();
                return (IFileSystemAccessService)service;
            });
```

However if possible, I think this is better:

```cs
        return serviceCollection
            .AddScoped<IFileSystemAccessServiceInProcess, FileSystemAccessServiceInProcess>()
            .AddScoped<IFileSystemAccessService, FileSystemAccessService>();
```

- During development accidentally changed `FileSystemAccessService` and `FileSystemAccessServiceInProcess` to `internal` because it needs to be consistent with `BaseFileSystemAccessService` (I changed all of them to `public` now). Though I thought it would be better (?) since we already provided the extension?

KristofferStrube commented on samples/KristofferStrube.Blazor.FileSystemAccess.WasmExample/wwwroot/content/custom-path/KristofferStrube.Blazor.FileSystemAccess.js:
```diff
@@ -0,0 +1,22 @@
+    var buffer = await blob.arrayBuffer();
+    var bytes = new Uint8Array(buffer);
+    return bytes;
+}
```
We can delete this file from the PR as we don't need it.

KristofferStrube commented on samples/KristofferStrube.Blazor.FileSystemAccess.WasmExample/Program.cs:
```diff
@@ -10,7 +10,14 @@
+builder.Services.AddFileSystemAccessServiceInProcess(options =>
+{
+    // The file at this path in this example is manually copied to wwwroot folder
+    // options.ScriptPath = $"/content/custom-path/{FileSystemAccessOptions.DefaultNamespace}.js";
```
The sample start of the path should not contain `"/content/"` as we changed that structure.

KristofferStrube reviewed (changes requested):
A lot of great changes to minimize repeated code. With very few changes then I think it is ready to Merge into Main and get a new Release out.

KristofferStrube commented:
Just to answer your previous list of comments.
- I had seen that problem before so I'm happy that you fixed it. I actually plan to combine the two different option types like `OpenFilePickerOptionsStartInWellKnownDirectory` and `OpenFilePickerOptionsStartInFileSystemHandle` into a single type using implicit converters like I have done [here](https://github.com/KristofferStrube/Blazor.ServiceWorker/blob/main/src/KristofferStrube.Blazor.ServiceWorker/Options/RequestInfo.cs) in another project. So the methods on the `IFileSystemAccessService` will be even slicker then.
- The reason for why `FileSystemAccessServiceInProcess` extended `FileSystemAccessService` was so that we could cast `FileSystemAccessServiceInProcess` to `FileSystemAccessService`, but I see that you achieved the same using the interfaces and I have tested that they are correctly fetched from DI so that is awesome!
- The reason I have kept the `FileSystemAccessService` and `FileSystemAccessServiceInProcess` public is so that people can potentially inject them however they want in their project. A well-known example of this is people who use Singleton services which would mean that they need all dependant services to also be injected as Singletons whereas our standard is a scoped service.
And I got your PR changes. No need to do anything other that pushing to your branch. :)

One other things that I didn't remember to add to my previous review.
- I don't think we need to add any of the methods in the `IFileSystemAccessServiceInProcess` as they have the same definition as what we new them up to.

datvm commented:
Awesome! I have pushed a small commit to fixed your last review items. As for the "extra" methods in `FileSystemAccessServiceInProcess`, let's just leave them there and fix it when we actually need to change it I guess 😅

KristofferStrube reviewed (approved):
Awesome Job!
